// noflo/components/WebService/CouchDB/OpenDatabase.js
// Generated by CoffeeScript 1.3.1
(function() {
  var OpenDatabase, couch, noflo,
    __hasProp = {}.hasOwnProperty,
    __extends = function(child, parent) { for (var key in parent) { if (__hasProp.call(parent, key)) child[key] = parent[key]; } function ctor() { this.constructor = child; } ctor.prototype = parent.prototype; child.prototype = new ctor; child.__super__ = parent.prototype; return child; };

  noflo = require("noflo");

  couch = require("couch-client");

  OpenDatabase = (function(_super) {

    __extends(OpenDatabase, _super);

    OpenDatabase.name = 'OpenDatabase';

    function OpenDatabase() {
      var _this = this;
      this.inPorts = {
        url: new noflo.Port()
      };
      this.outPorts = {
        connection: new noflo.ArrayPort()
      };
      this.inPorts.url.on("data", function(data) {
        var db;
        db = couch(data);
        return _this.createDatabase(db, _this.outPorts.connection);
      });
    }

    OpenDatabase.prototype.createDatabase = function(connection, outPort) {
      return connection.request("PUT", "/" + connection.uri.pathname, function(err, result) {
        if (err) {
          console.error(err);
        }
        outPort.send(connection);
        return outPort.disconnect();
      });
    };

    return OpenDatabase;

  })(noflo.Component);

  exports.getComponent = function() {
    return new OpenDatabase;
  };

}).call(this);
