// noflo/components/HTTP/BasicAuth.js
// Generated by CoffeeScript 1.3.1
(function() {
  var BasicAuth, connect, noflo,
    __hasProp = {}.hasOwnProperty,
    __extends = function(child, parent) { for (var key in parent) { if (__hasProp.call(parent, key)) child[key] = parent[key]; } function ctor() { this.constructor = child; } ctor.prototype = parent.prototype; child.prototype = new ctor; child.__super__ = parent.prototype; return child; };

  noflo = require("noflo");

  connect = require("connect");

  BasicAuth = (function(_super) {

    __extends(BasicAuth, _super);

    BasicAuth.name = 'BasicAuth';

    BasicAuth.prototype.description = "This component receives a HTTP request (req, res) combination on input, and runs the connect.basicAuth middleware for that";

    function BasicAuth() {
      var _this = this;
      this.request = null;
      this.inPorts = {
        "in": new noflo.Port()
      };
      this.outPorts = {
        out: new noflo.Port()
      };
      this.inPorts["in"].on("data", function(request) {
        return _this.request = request;
      });
      this.inPorts["in"].on("disconnect", function() {
        return connect.basicAuth(_this.authenticate)(_this.request.req, _this.request.res, function() {
          _this.outPorts.out.send(_this.request);
          _this.request = null;
          return _this.outPorts.out.disconnect();
        });
      });
    }

    BasicAuth.prototype.authenticate = function(login, password) {
      return login === "user" && password === "pass";
    };

    return BasicAuth;

  })(noflo.Component);

  exports.getComponent = function() {
    return new BasicAuth();
  };

}).call(this);
