// noflo/components/HTTP/SendResponse.js
// Generated by CoffeeScript 1.3.1
(function() {
  var SendResponse, noflo,
    __hasProp = {}.hasOwnProperty,
    __extends = function(child, parent) { for (var key in parent) { if (__hasProp.call(parent, key)) child[key] = parent[key]; } function ctor() { this.constructor = child; } ctor.prototype = parent.prototype; child.prototype = new ctor; child.__super__ = parent.prototype; return child; };

  noflo = require("noflo");

  SendResponse = (function(_super) {

    __extends(SendResponse, _super);

    SendResponse.name = 'SendResponse';

    SendResponse.prototype.description = "This component receives a HTTP request (req, res, next) combination on on input, and runs res.end(), sending the response to the user";

    function SendResponse() {
      var _this = this;
      this.request = null;
      this.inPorts = {
        "in": new noflo.Port()
      };
      this.outPorts = {};
      this.inPorts["in"].on("data", function(request) {
        return _this.request = request;
      });
      this.inPorts["in"].on("disconnect", function() {
        _this.request.res.end();
        return _this.request = null;
      });
    }

    return SendResponse;

  })(noflo.Component);

  exports.getComponent = function() {
    return new SendResponse();
  };

}).call(this);
