// noflo/components/CreateObject.js
// Generated by CoffeeScript 1.3.1
(function() {
  var CreateObject, noflo,
    __hasProp = {}.hasOwnProperty,
    __extends = function(child, parent) { for (var key in parent) { if (__hasProp.call(parent, key)) child[key] = parent[key]; } function ctor() { this.constructor = child; } ctor.prototype = parent.prototype; child.prototype = new ctor; child.__super__ = parent.prototype; return child; };

  noflo = require("noflo");

  CreateObject = (function(_super) {

    __extends(CreateObject, _super);

    CreateObject.name = 'CreateObject';

    function CreateObject() {
      var _this = this;
      this.inPorts = {
        start: new noflo.Port()
      };
      this.outPorts = {
        out: new noflo.Port()
      };
      this.inPorts.start.on("data", function() {
        _this.outPorts.out.send({});
        return _this.outPorts.out.disconnect();
      });
    }

    return CreateObject;

  })(noflo.Component);

  exports.getComponent = function() {
    return new CreateObject;
  };

}).call(this);
