// noflo/components/GetObjectKey.js
// Generated by CoffeeScript 1.3.1
(function() {
  var GetObjectKey, noflo,
    __hasProp = {}.hasOwnProperty,
    __extends = function(child, parent) { for (var key in parent) { if (__hasProp.call(parent, key)) child[key] = parent[key]; } function ctor() { this.constructor = child; } ctor.prototype = parent.prototype; child.prototype = new ctor; child.__super__ = parent.prototype; return child; };

  noflo = require("noflo");

  GetObjectKey = (function(_super) {

    __extends(GetObjectKey, _super);

    GetObjectKey.name = 'GetObjectKey';

    function GetObjectKey() {
      var _this = this;
      this.data = [];
      this.key = [];
      this.inPorts = {
        "in": new noflo.Port(),
        key: new noflo.ArrayPort()
      };
      this.outPorts = {
        out: new noflo.Port()
      };
      this.inPorts["in"].on("connect", function() {
        return _this.data = [];
      });
      this.inPorts["in"].on("begingroup", function(group) {
        return _this.outPorts.out.beginGroup(group);
      });
      this.inPorts["in"].on("data", function(data) {
        if (_this.key.length) {
          return _this.getKey(data);
        }
        return _this.data.push(data);
      });
      this.inPorts["in"].on("endgroup", function() {
        return _this.outPorts.out.endGroup();
      });
      this.inPorts["in"].on("disconnect", function() {
        var data, _i, _len, _ref;
        if (!_this.data.length) {
          _this.outPorts.out.disconnect();
          return;
        }
        if (!_this.key.length) {
          return;
        }
        _ref = _this.data;
        for (_i = 0, _len = _ref.length; _i < _len; _i++) {
          data = _ref[_i];
          _this.getKey(data);
        }
        return _this.outPorts.out.disconnect();
      });
      this.inPorts.key.on("data", function(data) {
        return _this.key.push(data);
      });
      this.inPorts.key.on("disconnect", function() {
        var data, _i, _len, _ref;
        if (!_this.data.length) {
          return;
        }
        _ref = _this.data;
        for (_i = 0, _len = _ref.length; _i < _len; _i++) {
          data = _ref[_i];
          _this.getKey(data);
        }
        _this.data = [];
        return _this.outPorts.out.disconnect();
      });
    }

    GetObjectKey.prototype.getKey = function(data) {
      var key, _i, _len, _ref, _results;
      if (!this.key.length) {
        throw "Key not defined";
      }
      if (typeof data !== "object") {
        throw "Data is not an object";
      }
      _ref = this.key;
      _results = [];
      for (_i = 0, _len = _ref.length; _i < _len; _i++) {
        key = _ref[_i];
        if (!data[key]) {
          continue;
        }
        if (this.key.length !== 1) {
          this.outPorts.out.beginGroup(key);
        }
        this.outPorts.out.send(data[key]);
        if (this.key.length !== 1) {
          _results.push(this.outPorts.out.endGroup());
        } else {
          _results.push(void 0);
        }
      }
      return _results;
    };

    return GetObjectKey;

  })(noflo.Component);

  exports.getComponent = function() {
    return new GetObjectKey;
  };

}).call(this);
