// noflo/components/GroupByPacket.js
// Generated by CoffeeScript 1.3.1
(function() {
  var GroupByPacket, noflo,
    __hasProp = {}.hasOwnProperty,
    __extends = function(child, parent) { for (var key in parent) { if (__hasProp.call(parent, key)) child[key] = parent[key]; } function ctor() { this.constructor = child; } ctor.prototype = parent.prototype; child.prototype = new ctor; child.__super__ = parent.prototype; return child; };

  noflo = require("noflo");

  GroupByPacket = (function(_super) {

    __extends(GroupByPacket, _super);

    GroupByPacket.name = 'GroupByPacket';

    function GroupByPacket() {
      var _this = this;
      this.packets = 0;
      this.inPorts = {
        "in": new noflo.Port
      };
      this.outPorts = {
        out: new noflo.Port
      };
      this.inPorts["in"].on("begingroup", function(group) {
        _this.outPorts.out.beginGroup(group);
        return _this.packets = 0;
      });
      this.inPorts["in"].on("data", function(data) {
        _this.outPorts.out.beginGroup(_this.packets);
        _this.outPorts.out.send(data);
        _this.outPorts.out.endGroup();
        return _this.packets++;
      });
      this.inPorts["in"].on("endgroup", function() {
        return _this.outPorts.out.endGroup();
      });
      this.inPorts["in"].on("disconnect", function() {
        _this.packets = 0;
        return _this.outPorts.out.disconnect();
      });
    }

    return GroupByPacket;

  })(noflo.Component);

  exports.getComponent = function() {
    return new GroupByPacket;
  };

}).call(this);
