// noflo/components/Kick.js
// Generated by CoffeeScript 1.3.1
(function() {
  var Kick, noflo,
    __hasProp = {}.hasOwnProperty,
    __extends = function(child, parent) { for (var key in parent) { if (__hasProp.call(parent, key)) child[key] = parent[key]; } function ctor() { this.constructor = child; } ctor.prototype = parent.prototype; child.prototype = new ctor; child.__super__ = parent.prototype; return child; };

  noflo = require("noflo");

  Kick = (function(_super) {

    __extends(Kick, _super);

    Kick.name = 'Kick';

    Kick.prototype.description = "This component generates a single packet and sends in to the output port. Mostly usable for debugging, but can also be useful for starting up networks.";

    function Kick() {
      var _this = this;
      this.inPorts = {
        "in": new noflo.Port()
      };
      this.outPorts = {
        out: new noflo.Port()
      };
      this.inPorts["in"].on("disconnect", function() {
        _this.outPorts.out.send(null);
        return _this.outPorts.out.disconnect();
      });
    }

    return Kick;

  })(noflo.Component);

  exports.getComponent = function() {
    return new Kick;
  };

}).call(this);
