// noflo/components/Merge.js
// Generated by CoffeeScript 1.3.1
(function() {
  var Merge, noflo,
    __hasProp = {}.hasOwnProperty,
    __extends = function(child, parent) { for (var key in parent) { if (__hasProp.call(parent, key)) child[key] = parent[key]; } function ctor() { this.constructor = child; } ctor.prototype = parent.prototype; child.prototype = new ctor; child.__super__ = parent.prototype; return child; };

  noflo = require("noflo");

  Merge = (function(_super) {

    __extends(Merge, _super);

    Merge.name = 'Merge';

    Merge.prototype.description = "This component receives data on multiple input ports and sends the same data out to the connected output port";

    function Merge() {
      var _this = this;
      this.inPorts = {
        "in": new noflo.ArrayPort()
      };
      this.outPorts = {
        out: new noflo.Port()
      };
      this.inPorts["in"].on("connect", function() {
        return _this.outPorts.out.connect();
      });
      this.inPorts["in"].on("begingroup", function(group) {
        return _this.outPorts.out.beginGroup(group);
      });
      this.inPorts["in"].on("data", function(data) {
        return _this.outPorts.out.send(data);
      });
      this.inPorts["in"].on("endgroup", function() {
        return _this.outPorts.out.endGroup();
      });
      this.inPorts["in"].on("disconnect", function() {
        return _this.outPorts.out.disconnect();
      });
    }

    return Merge;

  })(noflo.Component);

  exports.getComponent = function() {
    return new Merge;
  };

}).call(this);
