// noflo/components/PropertiesToObjects.js
// Generated by CoffeeScript 1.3.1
(function() {
  var PropertiesToObjects, noflo,
    __hasProp = {}.hasOwnProperty,
    __extends = function(child, parent) { for (var key in parent) { if (__hasProp.call(parent, key)) child[key] = parent[key]; } function ctor() { this.constructor = child; } ctor.prototype = parent.prototype; child.prototype = new ctor; child.__super__ = parent.prototype; return child; };

  noflo = require("noflo");

  PropertiesToObjects = (function(_super) {

    __extends(PropertiesToObjects, _super);

    PropertiesToObjects.name = 'PropertiesToObjects';

    function PropertiesToObjects() {
      var _this = this;
      this.inPorts = {
        "in": new noflo.Port()
      };
      this.outPorts = {
        out: new noflo.Port()
      };
      this.inPorts["in"].on("begingroup", function(group) {
        return _this.outPorts.out.beginGroup(group);
      });
      this.inPorts["in"].on("data", function(data) {
        return _this.convert(data);
      });
      this.inPorts["in"].on("endgroup", function() {
        return _this.outPorts.out.endGroup();
      });
      this.inPorts["in"].on("disconnect", function() {
        return _this.outPorts.out.disconnect();
      });
    }

    PropertiesToObjects.prototype.convert = function(object) {
      var data, id, objects, property, value;
      data = {};
      for (property in object) {
        objects = object[property];
        for (id in objects) {
          value = objects[id];
          if (!data[id]) {
            data[id] = {};
          }
          data[id][property] = value;
        }
      }
      return this.outPorts.out.send(data);
    };

    return PropertiesToObjects;

  })(noflo.Component);

  exports.getComponent = function() {
    return new PropertiesToObjects;
  };

}).call(this);
