// noflo/components/ReadEnv.js
// Generated by CoffeeScript 1.3.1
(function() {
  var ReadEnv, noflo,
    __hasProp = {}.hasOwnProperty,
    __extends = function(child, parent) { for (var key in parent) { if (__hasProp.call(parent, key)) child[key] = parent[key]; } function ctor() { this.constructor = child; } ctor.prototype = parent.prototype; child.prototype = new ctor; child.__super__ = parent.prototype; return child; };

  noflo = require('noflo');

  ReadEnv = (function(_super) {

    __extends(ReadEnv, _super);

    ReadEnv.name = 'ReadEnv';

    function ReadEnv() {
      var _this = this;
      this.inPorts = {
        key: new noflo.Port
      };
      this.outPorts = {
        out: new noflo.ArrayPort,
        error: new noflo.Port
      };
      this.inPorts.key.on('data', function(data) {
        if (process.env[data] !== void 0) {
          return _this.outPorts.out.send(process.env[data]);
        }
        if (_this.outPorts.error.isAttached()) {
          return _this.outPorts.error.send("No environment variable " + data + " set");
        }
      });
    }

    return ReadEnv;

  })(noflo.Component);

  exports.getComponent = function() {
    return new ReadEnv;
  };

}).call(this);
