// noflo/components/ReadFile.js
// Generated by CoffeeScript 1.3.1
(function() {
  var ReadFile, fs, noflo,
    __hasProp = {}.hasOwnProperty,
    __extends = function(child, parent) { for (var key in parent) { if (__hasProp.call(parent, key)) child[key] = parent[key]; } function ctor() { this.constructor = child; } ctor.prototype = parent.prototype; child.prototype = new ctor; child.__super__ = parent.prototype; return child; };

  fs = require("fs");

  noflo = require("noflo");

  ReadFile = (function(_super) {

    __extends(ReadFile, _super);

    ReadFile.name = 'ReadFile';

    function ReadFile() {
      var _this = this;
      this.inPorts = {
        source: new noflo.Port()
      };
      this.outPorts = {
        out: new noflo.Port(),
        error: new noflo.Port()
      };
      this.inPorts.source.on("data", function(data) {
        return _this.readFile(data);
      });
    }

    ReadFile.prototype.readFile = function(fileName) {
      var _this = this;
      return fs.readFile(fileName, "utf-8", function(err, data) {
        if (err) {
          _this.outPorts.error.send(err);
          return _this.outPorts.error.disconnect();
        }
        _this.outPorts.out.beginGroup(fileName);
        _this.outPorts.out.send(data);
        _this.outPorts.out.endGroup();
        return _this.outPorts.out.disconnect();
      });
    };

    return ReadFile;

  })(noflo.Component);

  exports.getComponent = function() {
    return new ReadFile();
  };

}).call(this);
