// noflo/components/Replace.js
// Generated by CoffeeScript 1.3.1
(function() {
  var Replace, noflo,
    __hasProp = {}.hasOwnProperty,
    __extends = function(child, parent) { for (var key in parent) { if (__hasProp.call(parent, key)) child[key] = parent[key]; } function ctor() { this.constructor = child; } ctor.prototype = parent.prototype; child.prototype = new ctor; child.__super__ = parent.prototype; return child; };

  noflo = require("noflo");

  Replace = (function(_super) {

    __extends(Replace, _super);

    Replace.name = 'Replace';

    function Replace() {
      var _this = this;
      this.pattern = null;
      this.replacement = "";
      this.inPorts = {
        "in": new noflo.Port(),
        pattern: new noflo.Port(),
        replacement: new noflo.Port()
      };
      this.outPorts = {
        out: new noflo.Port()
      };
      this.inPorts.pattern.on("data", function(data) {
        return _this.pattern = new RegExp(data, 'g');
      });
      this.inPorts.replacement.on("data", function(data) {
        return _this.replacement = data;
      });
      this.inPorts["in"].on("begingroup", function(group) {
        return _this.outPorts.out.beginGroup(group);
      });
      this.inPorts["in"].on("data", function(data) {
        var string;
        string = data;
        if (_this.pattern != null) {
          string = data.replace(_this.pattern, _this.replacement);
        }
        return _this.outPorts.out.send(string);
      });
      this.inPorts["in"].on("endgroup", function() {
        return _this.outPorts.out.endGroup();
      });
      this.inPorts["in"].on("disconnect", function() {
        return _this.outPorts.out.disconnect();
      });
    }

    return Replace;

  })(noflo.Component);

  exports.getComponent = function() {
    return new Replace;
  };

}).call(this);
