// noflo/components/UniquePacket.js
// Generated by CoffeeScript 1.3.1
(function() {
  var UniquePacket, noflo,
    __hasProp = {}.hasOwnProperty,
    __extends = function(child, parent) { for (var key in parent) { if (__hasProp.call(parent, key)) child[key] = parent[key]; } function ctor() { this.constructor = child; } ctor.prototype = parent.prototype; child.prototype = new ctor; child.__super__ = parent.prototype; return child; };

  noflo = require("noflo");

  UniquePacket = (function(_super) {

    __extends(UniquePacket, _super);

    UniquePacket.name = 'UniquePacket';

    function UniquePacket() {
      var _this = this;
      this.seen = {};
      this.inPorts = {
        "in": new noflo.Port()
      };
      this.outPorts = {
        out: new noflo.Port()
      };
      this.inPorts["in"].on("data", function(data) {
        if (_this.unique(data)) {
          return _this.outPorts.out.send(data);
        }
      });
      this.inPorts["in"].on("disconnect", function() {
        return _this.outPorts.out.disconnect();
      });
    }

    UniquePacket.prototype.unique = function(packet) {
      var stringed;
      stringed = JSON.stringify(packet);
      if (this.seen[stringed]) {
        return false;
      }
      this.seen[stringed] = true;
      return true;
    };

    return UniquePacket;

  })(noflo.Component);

  exports.getComponent = function() {
    return new UniquePacket;
  };

}).call(this);
