// noflo/components/UniqueArray.js
// Generated by CoffeeScript 1.3.1
(function() {
  var UniqueArray, noflo,
    __hasProp = {}.hasOwnProperty,
    __extends = function(child, parent) { for (var key in parent) { if (__hasProp.call(parent, key)) child[key] = parent[key]; } function ctor() { this.constructor = child; } ctor.prototype = parent.prototype; child.prototype = new ctor; child.__super__ = parent.prototype; return child; };

  noflo = require("noflo");

  UniqueArray = (function(_super) {

    __extends(UniqueArray, _super);

    UniqueArray.name = 'UniqueArray';

    function UniqueArray() {
      var _this = this;
      this.inPorts = {
        "in": new noflo.Port()
      };
      this.outPorts = {
        out: new noflo.Port()
      };
      this.inPorts["in"].on("data", function(data) {
        return _this.outPorts.out.send(_this.unique(data));
      });
      this.inPorts["in"].on("disconnect", function() {
        return _this.outPorts.out.disconnect();
      });
    }

    UniqueArray.prototype.unique = function(array) {
      var member, newArray, seen, _i, _len;
      seen = {};
      newArray = [];
      for (_i = 0, _len = array.length; _i < _len; _i++) {
        member = array[_i];
        seen[member] = member;
      }
      for (member in seen) {
        newArray.push(member);
      }
      return newArray;
    };

    return UniqueArray;

  })(noflo.Component);

  exports.getComponent = function() {
    return new UniqueArray;
  };

}).call(this);
