// noflo/lib/InternalSocket.js
// Generated by CoffeeScript 1.3.1
(function() {
  var InternalSocket, events,
    __hasProp = {}.hasOwnProperty,
    __extends = function(child, parent) { for (var key in parent) { if (__hasProp.call(parent, key)) child[key] = parent[key]; } function ctor() { this.constructor = child; } ctor.prototype = parent.prototype; child.prototype = new ctor; child.__super__ = parent.prototype; return child; };

  events = require("events");

  InternalSocket = (function(_super) {

    __extends(InternalSocket, _super);

    InternalSocket.name = 'InternalSocket';

    function InternalSocket() {
      this.connected = false;
      this.groups = [];
    }

    InternalSocket.prototype.connect = function() {
      this.connected = true;
      return this.emit('connect', this);
    };

    InternalSocket.prototype.disconnect = function() {
      this.connected = false;
      return this.emit('disconnect', this);
    };

    InternalSocket.prototype.isConnected = function() {
      return this.connected;
    };

    InternalSocket.prototype.send = function(data) {
      return this.emit('data', data);
    };

    InternalSocket.prototype.beginGroup = function(group) {
      this.groups.push(group);
      return this.emit('begingroup', group);
    };

    InternalSocket.prototype.endGroup = function() {
      return this.emit('endgroup', this.groups.pop());
    };

    InternalSocket.prototype.getId = function() {
      if (this.from && !this.to) {
        return "" + this.from.process.id + ":" + this.from.port + " -> ANON";
      }
      if (!this.from) {
        return "DATA -> " + this.to.process.id + ":" + this.to.port;
      }
      return "" + this.from.process.id + ":" + this.from.port + " -> " + this.to.process.id + ":" + this.to.port;
    };

    return InternalSocket;

  })(events.EventEmitter);

  exports.InternalSocket = InternalSocket;

  exports.createSocket = function() {
    return new InternalSocket;
  };

}).call(this);
