// noflo/lib/Port.js
// Generated by CoffeeScript 1.3.1
(function() {
  var Port, events,
    __hasProp = {}.hasOwnProperty,
    __extends = function(child, parent) { for (var key in parent) { if (__hasProp.call(parent, key)) child[key] = parent[key]; } function ctor() { this.constructor = child; } ctor.prototype = parent.prototype; child.prototype = new ctor; child.__super__ = parent.prototype; return child; };

  events = require("events");

  Port = (function(_super) {

    __extends(Port, _super);

    Port.name = 'Port';

    function Port(name) {
      this.name = name;
      this.socket = null;
      this.from = null;
    }

    Port.prototype.attach = function(socket) {
      if (this.socket) {
        throw new Error("" + this.name + ": Socket already attached " + (this.socket.getId()) + " - " + (socket.getId()));
      }
      this.socket = socket;
      return this.attachSocket(socket);
    };

    Port.prototype.attachSocket = function(socket) {
      var _this = this;
      this.emit("attach", socket);
      this.from = socket.from;
      socket.setMaxListeners(0);
      socket.on("connect", function() {
        return _this.emit("connect", socket);
      });
      socket.on("begingroup", function(group) {
        return _this.emit("begingroup", group);
      });
      socket.on("data", function(data) {
        return _this.emit("data", data);
      });
      socket.on("endgroup", function(group) {
        return _this.emit("endgroup", group);
      });
      return socket.on("disconnect", function() {
        return _this.emit("disconnect", socket);
      });
    };

    Port.prototype.connect = function() {
      if (!this.socket) {
        throw new Error("No connection available");
      }
      return this.socket.connect();
    };

    Port.prototype.beginGroup = function(group) {
      var _this = this;
      if (!this.socket) {
        throw new Error("No connection available");
      }
      if (this.isConnected()) {
        return this.socket.beginGroup(group);
      }
      this.socket.once("connect", function() {
        return _this.socket.beginGroup(group);
      });
      return this.socket.connect();
    };

    Port.prototype.send = function(data) {
      var _this = this;
      if (!this.socket) {
        throw new Error("No connection available");
      }
      if (this.isConnected()) {
        return this.socket.send(data);
      }
      this.socket.once("connect", function() {
        return _this.socket.send(data);
      });
      return this.socket.connect();
    };

    Port.prototype.endGroup = function() {
      if (!this.socket) {
        throw new Error("No connection available");
      }
      return this.socket.endGroup();
    };

    Port.prototype.disconnect = function() {
      if (!this.socket) {
        return;
      }
      return this.socket.disconnect();
    };

    Port.prototype.detach = function(socket) {
      this.emit("detach", this.socket);
      this.from = null;
      return this.socket = null;
    };

    Port.prototype.isConnected = function() {
      if (!this.socket) {
        return false;
      }
      return this.socket.isConnected();
    };

    Port.prototype.isAttached = function() {
      return this.socket !== null;
    };

    return Port;

  })(events.EventEmitter);

  exports.Port = Port;

}).call(this);
