// noflo/components/WebService/Basecamp/ReadProjects.js
// Generated by CoffeeScript 1.3.1
(function() {
  var ReadProjects, base, noflo,
    __hasProp = {}.hasOwnProperty,
    __extends = function(child, parent) { for (var key in parent) { if (__hasProp.call(parent, key)) child[key] = parent[key]; } function ctor() { this.constructor = child; } ctor.prototype = parent.prototype; child.prototype = new ctor; child.__super__ = parent.prototype; return child; };

  noflo = require("noflo");

  base = require("./BasecampComponent");

  ReadProjects = (function(_super) {

    __extends(ReadProjects, _super);

    ReadProjects.name = 'ReadProjects';

    function ReadProjects() {
      var _this = this;
      this.basePortSetup();
      this.outPorts = {
        out: new noflo.Port()
      };
      this.inPorts.apikey.on("disconnect", function() {
        if (_this.hostname) {
          return _this.readProject();
        }
      });
      this.inPorts.hostname.on("disconnect", function() {
        if (_this.apikey) {
          return _this.readProject();
        }
      });
    }

    ReadProjects.prototype.readProject = function() {
      var _this = this;
      return this.get("/projects.xml", function(data) {
        return _this.parseProject(data);
      });
    };

    ReadProjects.prototype.parseProject = function(data) {
      var id, target;
      target = this.outPorts.out;
      id = "https://" + this.hostname + "/";
      return this.parse(data, function(parsed) {
        var project, _i, _len, _ref;
        target.beginGroup(id);
        _ref = parsed.project;
        for (_i = 0, _len = _ref.length; _i < _len; _i++) {
          project = _ref[_i];
          target.send(project);
        }
        target.endGroup();
        return target.disconnect();
      });
    };

    return ReadProjects;

  })(base.BasecampComponent);

  exports.getComponent = function() {
    return new ReadProjects;
  };

}).call(this);
