// noflo/components/WebService/Basecamp/ReadTaskLists.js
// Generated by CoffeeScript 1.3.1
(function() {
  var ReadTaskLists, base, noflo,
    __hasProp = {}.hasOwnProperty,
    __extends = function(child, parent) { for (var key in parent) { if (__hasProp.call(parent, key)) child[key] = parent[key]; } function ctor() { this.constructor = child; } ctor.prototype = parent.prototype; child.prototype = new ctor; child.__super__ = parent.prototype; return child; };

  noflo = require("noflo");

  base = require("./BasecampComponent");

  ReadTaskLists = (function(_super) {

    __extends(ReadTaskLists, _super);

    ReadTaskLists.name = 'ReadTaskLists';

    function ReadTaskLists() {
      var _this = this;
      this.project = null;
      this.basePortSetup();
      this.inPorts.project = new noflo.Port();
      this.outPorts = {
        out: new noflo.Port()
      };
      this.inPorts.project.on("data", function(data) {
        _this.project = data;
        if (_this.hostname && _this.apikey) {
          return _this.readTaskLists();
        }
      });
    }

    ReadTaskLists.prototype.readTaskListsAll = function() {
      var _this = this;
      return this.get("/todo_lists.xml", function(data) {
        return _this.parseTaskLists(data);
      });
    };

    ReadTaskLists.prototype.readTaskLists = function() {
      var _this = this;
      if (!this.project) {
        return this.readTaskListsAll;
      }
      return this.get("/projects/" + this.project + "/todo_lists.xml", function(data) {
        return _this.parseTaskLists(data);
      });
    };

    ReadTaskLists.prototype.parseTaskLists = function(data) {
      var id, target;
      target = this.outPorts.out;
      id = "https://" + this.hostname + "/";
      return this.parse(data, function(parsed) {
        var taskList, _i, _len, _ref;
        if (!parsed['todo-list']) {
          return;
        }
        target.beginGroup(id);
        if (toString.call(parsed['todo-list']) !== '[object Array]') {
          target.send(parsed['todo-list']);
          target.endGroup();
          return target.disconnect();
        }
        _ref = parsed['todo-list'];
        for (_i = 0, _len = _ref.length; _i < _len; _i++) {
          taskList = _ref[_i];
          target.send(taskList);
        }
        target.endGroup();
        return target.disconnect();
      });
    };

    return ReadTaskLists;

  })(base.BasecampComponent);

  exports.getComponent = function() {
    return new ReadTaskLists;
  };

}).call(this);
