// noflo/components/WebService/Basecamp/ReadTasks.js
// Generated by CoffeeScript 1.3.1
(function() {
  var ReadTasks, base, noflo,
    __hasProp = {}.hasOwnProperty,
    __extends = function(child, parent) { for (var key in parent) { if (__hasProp.call(parent, key)) child[key] = parent[key]; } function ctor() { this.constructor = child; } ctor.prototype = parent.prototype; child.prototype = new ctor; child.__super__ = parent.prototype; return child; };

  noflo = require("noflo");

  base = require("./BasecampComponent");

  ReadTasks = (function(_super) {

    __extends(ReadTasks, _super);

    ReadTasks.name = 'ReadTasks';

    function ReadTasks() {
      var _this = this;
      this.tasklist = null;
      this.basePortSetup();
      this.inPorts.tasklist = new noflo.Port();
      this.outPorts = {
        out: new noflo.Port()
      };
      this.inPorts.tasklist.on("data", function(data) {
        _this.tasklist = data;
        if (_this.hostname && _this.apikey) {
          return _this.readTasks();
        }
      });
    }

    ReadTasks.prototype.readTasks = function() {
      var _this = this;
      return this.get("/todo_lists/" + this.tasklist + "/todo_items.xml", function(data) {
        return _this.parseTasks(data);
      });
    };

    ReadTasks.prototype.parseTasks = function(data) {
      var id, target;
      target = this.outPorts.out;
      id = "https://" + this.hostname + "/";
      return this.parse(data, function(parsed) {
        var task, _i, _len, _ref;
        if (!parsed['todo-item']) {
          return;
        }
        target.beginGroup(id);
        if (toString.call(parsed['todo-item']) !== '[object Array]') {
          target.send(parsed['todo-item']);
          target.endGroup();
          return target.disconnect();
        }
        _ref = parsed['todo-item'];
        for (_i = 0, _len = _ref.length; _i < _len; _i++) {
          task = _ref[_i];
          target.send(task);
        }
        target.endGroup();
        return target.disconnect();
      });
    };

    return ReadTasks;

  })(base.BasecampComponent);

  exports.getComponent = function() {
    return new ReadTasks;
  };

}).call(this);
