// noflo/components/WebService/Basecamp/ReadHours.js
// Generated by CoffeeScript 1.3.1
(function() {
  var ReadHours, base, noflo,
    __hasProp = {}.hasOwnProperty,
    __extends = function(child, parent) { for (var key in parent) { if (__hasProp.call(parent, key)) child[key] = parent[key]; } function ctor() { this.constructor = child; } ctor.prototype = parent.prototype; child.prototype = new ctor; child.__super__ = parent.prototype; return child; };

  noflo = require("noflo");

  base = require("./BasecampComponent");

  ReadHours = (function(_super) {

    __extends(ReadHours, _super);

    ReadHours.name = 'ReadHours';

    function ReadHours() {
      var _this = this;
      this.start = false;
      this.basePortSetup();
      this.inPorts.start = new noflo.Port();
      this.outPorts = {
        out: new noflo.Port()
      };
      this.inPorts.start.on("data", function() {
        _this.start = true;
        if (_this.hostname && _this.apikey) {
          return _this.readHours();
        }
      });
      this.inPorts.apikey.on("disconnect", function() {
        if (_this.hostname && _this.start) {
          return _this.readHours();
        }
      });
      this.inPorts.hostname.on("disconnect", function() {
        if (_this.apikey && _this.start) {
          return _this.readHours();
        }
      });
    }

    ReadHours.prototype.readHours = function() {
      var _this = this;
      return this.get("/time_entries/report.xml", function(data) {
        return _this.parseHours(data);
      });
    };

    ReadHours.prototype.parseHours = function(data) {
      var id, target;
      target = this.outPorts.out;
      id = "https://" + this.hostname + "/";
      return this.parse(data, function(parsed) {
        var entry, _i, _len, _ref;
        target.beginGroup(id);
        _ref = parsed['time-entry'];
        for (_i = 0, _len = _ref.length; _i < _len; _i++) {
          entry = _ref[_i];
          target.send(entry);
        }
        target.endGroup();
        return target.disconnect();
      });
    };

    return ReadHours;

  })(base.BasecampComponent);

  exports.getComponent = function() {
    return new ReadHours;
  };

}).call(this);
