// noflo/components/WebService/Basecamp/ReadProjectHours.js
// Generated by CoffeeScript 1.3.1
(function() {
  var ReadProjectHours, base, noflo,
    __hasProp = {}.hasOwnProperty,
    __extends = function(child, parent) { for (var key in parent) { if (__hasProp.call(parent, key)) child[key] = parent[key]; } function ctor() { this.constructor = child; } ctor.prototype = parent.prototype; child.prototype = new ctor; child.__super__ = parent.prototype; return child; };

  noflo = require("noflo");

  base = require("./BasecampComponent");

  ReadProjectHours = (function(_super) {

    __extends(ReadProjectHours, _super);

    ReadProjectHours.name = 'ReadProjectHours';

    function ReadProjectHours() {
      var _this = this;
      this.project = null;
      this.basePortSetup();
      this.inPorts.project = new noflo.Port();
      this.outPorts = {
        out: new noflo.Port()
      };
      this.inPorts.project.on("data", function(data) {
        return _this.project = data;
      });
      this.inPorts.project.on("disconnect", function() {
        if (_this.hostname && _this.apikey) {
          return _this.readHours();
        }
      });
      this.inPorts.apikey.on("disconnect", function() {
        if (_this.hostname && _this.project) {
          return _this.readHours();
        }
      });
      this.inPorts.hostname.on("disconnect", function() {
        if (_this.apikey && _this.project) {
          return _this.readHours();
        }
      });
    }

    ReadProjectHours.prototype.readHours = function() {
      var _this = this;
      return this.get("/projects/" + this.project + "/time_entries.xml", function(data) {
        return _this.parseHours(data);
      });
    };

    ReadProjectHours.prototype.parseHours = function(data) {
      var id, target;
      target = this.outPorts.out;
      id = "https://" + this.hostname + "/";
      return this.parse(data, function(parsed) {
        var entry, _i, _len, _ref;
        target.beginGroup(id);
        _ref = parsed['time-entry'];
        for (_i = 0, _len = _ref.length; _i < _len; _i++) {
          entry = _ref[_i];
          target.send(entry);
        }
        target.endGroup();
        return target.disconnect();
      });
    };

    return ReadProjectHours;

  })(base.BasecampComponent);

  exports.getComponent = function() {
    return new ReadProjectHours;
  };

}).call(this);
